src/vm-memory/src/lib.rs

```diff
@@ -5,18 +5,329 @@
+    GuestUsize, MemoryRegionAddress,
+};
 
-// Export local backend implementation.
-pub use mmap::{GuestMemoryMmap, GuestRegionMmap};
+use std::io::Error as IoError;
+use std::os::unix::io::AsRawFd;
 
-// Re-export only what is needed in Firecracker.
-pub use vm_memory_upstream::{
-    address, Address, ByteValued, Bytes, Error, FileOffset, GuestAddress, GuestMemory,
-    GuestMemoryError, GuestMemoryRegion, MemoryRegionAddress, MmapRegion,
+use vm_memory_upstream::bitmap::AtomicBitmap;
+use vm_memory_upstream::mmap::{check_file_offset, NewBitmap};
+use vm_memory_upstream::{
+    GuestMemoryMmap as UpstreamGuestMemoryMmap, GuestRegionMmap as UpstreamGuestRegionMmap,
+    MmapRegion as UpstreamMmapRegion,
 };
+
+pub type GuestMemoryMmap = UpstreamGuestMemoryMmap<Option<AtomicBitmap>>;
+pub type GuestRegionMmap = UpstreamGuestRegionMmap<Option<AtomicBitmap>>;
+pub type MmapRegion = UpstreamMmapRegion<Option<AtomicBitmap>>;
+
+const GUARD_PAGE_COUNT: usize = 2;
+
+/// Build a `MmapRegion` surrounded by guard pages.
+///
+/// Initially, we map a `PROT_NONE` guard region of size: `size` + (GUARD_PAGE_COUNT * page_size).
+/// The guard region is mapped with `PROT_NONE`, so that any access to this region will cause
+/// a SIGSEGV.
+///
+/// The actual accessible region is going to be nested in the larger guard region.
+/// This is done by mapping over the guard region, starting at an address of
+/// `guard_region_addr + (GUARD_PAGE_COUNT / 2 * page_size)`.
+/// This results in a border of `GUARD_PAGE_COUNT / 2` pages on either side of the region, which
+/// acts as a safety net for accessing out-of-bounds addresses that are not allocated for the
+/// guest's memory.
+fn build_guarded_region(
+    maybe_file_offset: Option<FileOffset>,
+    size: usize,
+    prot: i32,
+    flags: i32,
+    track_dirty_pages: bool,
+) -> Result<MmapRegion, MmapRegionError> {
+    let page_size = unsafe { libc::sysconf(libc::_SC_PAGESIZE) as usize };
+    // Create the guarded range size (received size + X pages),
+    // where X is defined as a constant GUARD_PAGE_COUNT.
+    let guarded_size = size + GUARD_PAGE_COUNT * page_size;
+
+    // Map the guarded range to PROT_NONE
+    let guard_addr = unsafe {
+        libc::mmap(
+            std::ptr::null_mut(),
+            guarded_size,
+            libc::PROT_NONE,
+            libc::MAP_ANONYMOUS | libc::MAP_PRIVATE | libc::MAP_NORESERVE,
+            -1,
+            0,
+        )
+    };
+
+    if guard_addr == libc::MAP_FAILED {
+        return Err(MmapRegionError::Mmap(IoError::last_os_error()));
+    }
+
+    let (fd, offset) = match maybe_file_offset {
+        Some(ref file_offset) => {
+            check_file_offset(file_offset, size)?;
+            (file_offset.file().as_raw_fd(), file_offset.start())
+        }
+        None => (-1, 0),
+    };
+
+    let region_start_addr = guard_addr as usize + page_size * (GUARD_PAGE_COUNT / 2);
```

Comment:
This logic could be problematic if we set an odd value for `GUARD_PAGE_COUNT`. Maybe it would be better to have `GUARD_PAGE_COUNT_BEFORE`, `GUARD_PAGE_COUNT_AFTER` in order to make everything clear. Or we could have just `GUARD_PAGE_COUNT`, but have it represent the number of guard pages for each section (before, after) instead of the total number of guard pages. So in this case `GUARD_PAGE_COUNT` would be `1` and we would change the way we compute `region_start_addr` and `guarded_size`. Would it make sense ?

Reply:
good point. I'll fix this